source/common/logger/ur_sinks.hpp

```diff
@@ -0,0 +1,178 @@
+                    throw std::runtime_error("Only empty braces are allowed!");
+                } else {
+                    *ostream << arg;
+                    fmt++;
+                }
+            } else if (*fmt == '}') {
+                if (*(++fmt) == '}') {
+                    *ostream << *fmt++;
+                } else {
+                    throw std::runtime_error("Closing curly brace not escaped!");
+                }
+            }
+        }
+    }
+
+    template <typename Arg, typename... Args>
+    void format(const char *fmt, Arg &&arg, Args &&...args) {
+        bool arg_printed = false;
+        while (!arg_printed) {
+            while (*fmt != '{' && *fmt != '}' && *fmt != '\0') {
+                *ostream << *fmt++;
+            }
+
+            if (*fmt == '{') {
+                if (*(++fmt) == '{') {
+                    *ostream << *fmt++;
+                } else if (*fmt != '}') {
+                    throw std::runtime_error("Only empty braces are allowed!");
+                } else {
+                    *ostream << arg;
+                    arg_printed = true;
+                }
+            } else if (*fmt == '}') {
+                if (*(++fmt) == '}') {
+                    *ostream << *fmt++;
+                } else {
+                    throw std::runtime_error("Closing curly brace not escaped!");
+                }
+            }
+        }
+
+        format(++fmt, std::forward<Args &&>(args)...);
+    }
+};
+
+class StdoutSink : public Sink {
+  public:
+    StdoutSink() { this->ostream = &std::cout; }
+
+    ~StdoutSink() = default;
+};
+
+class StderrSink : public Sink {
+  public:
+    StderrSink() { this->ostream = &std::cerr; }
+
+    ~StderrSink() = default;
+};
+
+class FileSink : public Sink {
+  public:
+    FileSink(std::string file_path) {
+        ofstream = std::ofstream(file_path, std::ofstream::out);
+        if (ofstream.rdstate() != std::ofstream::goodbit) {
+            throw std::invalid_argument(
+                std::string("Failure while opening log file: ") + file_path +
+                std::string(" Check if given path exists."));
+        }
+        this->ostream = &ofstream;
+    }
+
+    ~FileSink() { ofstream.close(); }
+
+  private:
+    std::ofstream ofstream;
+};
+
+auto sink_from_str(std::string name) {
+    struct sink_name {
+        std::string name;
+        std::unique_ptr<Sink> sink;
+    };
+
+    sink_name sink_names[] = {
+        {"stdout", std::make_unique<logger::StdoutSink>()},
+        {"stderr", std::make_unique<logger::StderrSink>()}};
+
+    for (auto &item : sink_names) {
+        if (item.name.compare(name) == 0) {
+            return std::move(item.sink);
+        }
+    }
+    throw std::invalid_argument(
+        std::string("Parsing error: no valid sink for string '") + name +
+        std::string("'.") +
+        std::string("\nValid sink names are: stdout, stderr"));
+}
+
+auto sink_from_str(std::string name, std::string file_path) {
```

Comment:
I think it would be simpler if you just had single function that always accepted two arguments, with the file path being optional.

Reply:
Done